sdk/identity/identity/src/msal/browserFlows/msalImplicit.ts

```diff
@@ -0,0 +1,191 @@
+
+/**
+ * Uses MSAL directly for browser authentication,
+ * which uses the [Implicit Grant Flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow)
+ * @internal
+ */
+export class MSALImplicit extends MsalBrowser {
+  protected app: msal.UserAgentApplication;
+
+  /**
+   * Sets up an MSAL object based on the given parameters.
+   * MSAL with Implicit Grant is not compatible with the account workflow we use from the MSAL Auth Code flow.
+   * In this case, any `authenticationRecord` received will be ignored.
+   * @param options - Parameters necessary and otherwise used to create the MSAL object.
+   */
+  constructor(options: MsalBrowserFlowOptions) {
+    super(options);
+
+    this.msalConfig.cache = {
+      cacheLocation: "localStorage",
+      storeAuthStateInCookie: true // Set to true to improve the experience on IE11 and Edge.
+    };
+
+    // Preparing the MSAL application.
+    this.app = new msal.UserAgentApplication(this.msalConfig as msal.Configuration);
+  }
+
+  /**
+   * Formats an MSAL 1 account into an `AuthenticationRecord`.
+   * @param account - The account in the shape defined by MSAL.
+   */
+  private handleAccount(account: msal.Account): AuthenticationRecord {
+    return {
+      homeAccountId: account.homeAccountIdentifier,
+      environment: account.environment,
+      tenantId: this.tenantId,
+      localAccountId: account.accountIdentifier,
+      username: account.environment
+    };
+  }
+
+  /**
+   * Loads the account based on the result of the authentication.
+   * If no result was received, tries to load the account from the cache.
+   * @param result - Result object received from MSAL.
+   */
+  private async handleBrowserResult(
+    result?: msal.AuthResponse
+  ): Promise<AuthenticationRecord | undefined> {
+    if (result?.account) {
+      this.logger.info(`Authentication successful.`);
+      this.account = this.handleAccount(result?.account);
+      return this.account;
+    }
+    await this.getActiveAccount();
+    if (this.account) {
+      return this.account;
+    }
+    return;
+  }
+
+  private redirectPromise: Promise<AuthenticationRecord | undefined> | undefined;
+
+  /**
+   * Attempts to handle a redirection request the least amount of times possible.
+   */
+  public async handleRedirect(): Promise<AuthenticationRecord | undefined> {
+    if (this.account) {
+      return this.account;
+    }
+    if (this.redirectPromise) {
+      return this.redirectPromise;
+    }
+    if (!self.location.hash) {
+      return;
+    }
+    this.redirectPromise = new Promise((resolve, reject) => {
+      this.app.handleRedirectCallback(async (result) => {
+        if (!result?.account) {
+          const errorMessage = `Authentication failed. No redirect result.`;
+          this.logger.info(errorMessage);
+          reject(new Error(errorMessage));
+        }
+        await this.handleBrowserResult(result!);
```

Comment:
if this await throws, `redirectPromise` will never resolve. Recommend avoiding using `async` functions for the promise constructor

Reply:
redirectPromise will never resolve, since it redirects. I think it's better if we await, since this should not return!

Reply:
So nobody will be stuck waiting on redirectPromise? What's the point of having it then? That it just blocks until the redirect happens and the page unloads?